---
title: "Final Project"
output: pdf_document
---

```{r setup, include=FALSE}
library(TCGAbiolinks)
library(tidyr)
library(dplyr)
library(plyr)
library(ggplot2)
library(SummarizedExperiment)
library(DescTools)
library(qqman)
library(minfi)
library(purrr)
```

## Introduction to Data and Background:

According to the Mayo Clinic, Renal Cell Carcinoma is the most common kidney cancer and the incidence of this cancer is increasing in the population [1](). One of the key risk factors attributed to the onset of Renal Cell Carcinoma is familial history of the disease or transmission through genetics [1](). Furthermore, there is evidence to show that kidney cancers begin when mutations form in some kidney cells' DNA which causes the cells to grow and divide very rapidly [1](). Some other studies have shown that DNA Methylation can play a role in gene expression which can in turn lead to complex diseases such as cancer[2](https://pubmed.ncbi.nlm.nih.gov/15542813/). This can specifically be seen for Renal Cell Carcinoma as well as a previous study done by Lasseigne et al showed that there is a difference in the methylation profiles between patients affected by Renal Cell Carcinoma and those that are not in several genes [3](https://pubmed.ncbi.nlm.nih.gov/25472429/). For this reason, in this analysis we will study the relationship between DNA Methylation and Renal Cell Carcinoma through the use of data gathered from the TCGA database. The data lists a set of 325 Cases and 160 Controls as well as their methylation levels. 160 of the Case samples are paired with the Control Samples meaning that the Control Sample for these 160 patients is a tissue sample that is adjacent to the tumor sample collected from the same patient. Furthermore, 165 of the Case samples are Independent samples. Studying the relationship between DNA Methylation and Renal Cell Carcinoma can help us create targeted therapies, help us to understand how this disease occurs, and help us uncover which CpG sites or regions in the genome are correlated with Renal Cell Carcinoma the most.

## Methods:

### Data Cleaning and Preprocessing:

### Paired and Independent T-Tests:

To analyze the relationship between DNA Methylation and Renal Cell Carcinoma we will first perform an analysis to find differentially methylated regions (DMR) between the two groups. To do this we will perform two sets of different t tests, stratifying by paired and unpaired samples. For the 159 paired Case/Control samples we will conduct a paired two sided t test for each CpG site in our data. For the other 163 Case Samples, we will conduct a two sample two sided t test comparing them with the 160 Control Samples for each CpG site in our data. Our null hypothesis for these tests is that there is no difference in methylation between cases and controls and our alternate hypothesis is that there is a significant difference in methylation between cases and controls. We will use the standard significance level of 0.05 to determine significance for both the paired and two sample t tests we are doing. We will be performing multiple comparison adjustments on this result as we are comparing hundreds of thousands of CpG sites so not adjusting for multiple comparisons could lead to large errors in our analysis. The method we will use to correct for multiple comparisons is FDR since this allows us to have a better power than when using other methods such as Bonferroni, especially since our sample size is extremely large. 

### Plots:

To visualize our results effectively we will do volcano plots and manhattan plots for both our paired t test and independent t test results. The manhattan plots will allow us to visualize which regions of the genome have the most significant difference in methylation between Cases and Controls according to our t tests and the volcano plots will allow us to analyze if the CpG sites that are significant and have the largest change in methylation between cases and controls, as these CpG sites may be the most biologically relevant ones.  


## Data Download, Cleaning, and Processing:

# Query 
```{r}
# Query raw methylation 450K array from TCGA
query <- GDCquery(project = "TCGA-KIRC",
                      data.category = "Raw microarray data",
                      data.type = "Raw intensities", 
                      experimental.strategy = "Methylation array", 
                      legacy = TRUE,
                      file.type = ".idat",
                      platform = "Illumina Human Methylation 450")
clinical <- GDCquery_clinic(project = "TCGA-KIRC")
# Download IDAT files from TCGA
tryCatch(GDCdownload(query, method = "api", files.per.chunk = 20),
             error = function(e) GDCdownload(query, method = "client"))
# Move to current work dir
for(file in dir(".",pattern = ".idat", recursive = T)){
  TCGAbiolinks:::move(file,basename(file))
}
```

# Preprocess
Reference: https://bioconductor.org/help/course-materials/2014/BioC2014/minfi_BioC2014.pdf
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4916993/
https://www.nature.com/articles/bjc2013496
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5302158/
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3592906/
```{r}
# Reformat to a target data frame
results <- getResults(query) %>%
  group_by(cases) %>%
  mutate(Basename=substr(file_name, 1, 17)) %>%
  dplyr::select(cases, state, sample_type, cases.submitter_id, sample.submitter_id, Basename) %>% unique() %>%
  left_join(clinical %>%
              select(submitter_id, gender),
            by=c("cases.submitter_id"="submitter_id"))
results$sample_type[results$sample_type=="Additional - New Primary"] <- "Primary Tumor"

# rm(query)
  
# Read raw data
#setwd("/users/13038/downloads/Stat_Gen_Final_Project")
RGSets <- read.metharray.exp(targets=results)
manifest <- minfi::getManifest(RGSets)
manifest
head(getProbeInfo(manifest))
# rm(RGSets)

# Quality Control for Samples
MSets <- preprocessRaw(RGSets)
qc <- getQC(MSets)
head(qc)
plotQC(qc)
rm(MSets)

detP <- detectionP(RGSets)
detP[1:5,1:5]
keep <- colMeans(detP) < 0.01 # Remove sample with detection p-value > 0.01
table(keep) # No one is removed

# The 402nd sample seems to be bad from the QCplot, remove it
RGSets <- RGSets[,-402]
results <- results[-402,]
detP <- detP[,-402]

# Normalization (preprocessFunnorm: useful for can studies)
GRSets_norm <- preprocessFunnorm(RGSets)
plotSex(GRSets_norm)
# Some predicted sexes are wrong
GRSets_norm@colData$gender[GRSets_norm@colData$gender=="female"] <- "F"
GRSets_norm@colData$gender[GRSets_norm@colData$gender=="male"] <- "M"
GRSets_norm@colData$predictedSex <- GRSets_norm@colData$gender
plotSex(GRSets_norm)


# Beta densities for unnormalized & normalized data
par(mfrow=c(1,2))
densityPlot(RGSets, sampGroups = results$sample_type, main="Raw", legend=F)
legend("top", legend = levels(factor(results$sample_type)),
       text.col=RColorBrewer::brewer.pal(8,"Dark2"), cex = 0.6)
densityPlot(getBeta(GRSets_norm), sampGroups = results$sample_type, main="Normalized", legend=F)
legend("top", legend = levels(factor(results$sample_type)),
       text.col=RColorBrewer::brewer.pal(8,"Dark2"), cex = 0.6)

# Quality Control for probes
detP <- detP[match(featureNames(GRSets_norm),rownames(detP)),] # Ensure probes are in the same order
keep <- rowSums(detP < 0.01) == ncol(GRSets_norm) # Remove probes with one or more detection p-value > 0.01
table(keep)
GRSets_norm_filter <- GRSets_norm[keep,]

ann450k <-  getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
keep <- !(featureNames(GRSets_norm_filter) %in% 
            ann450k$Name[ann450k$chr %in% c("chrX", "chrY")])
table(keep) 
GRSets_norm_filter <- GRSets_norm_filter[keep,] # Remove probes on the sex chromosomes

GRSets_norm_filter <- dropLociWithSnps(GRSets_norm_filter) # Remove probes with SNPs at CpG sites

xReactiveProbes <- read.csv("/users/13038/downloads/Stat_Gen_Final_Project/data/48639-non-specific-probes-Illumina450k.csv")
keep <- !(featureNames(GRSets_norm_filter) %in% xReactiveProbes$TargetID)
table(keep)
GRSets_norm_filter <- GRSets_norm_filter[keep,] # Remove cross reactive probes

# Get M values and beta values
mVals <- getM(GRSets_norm_filter)
bVals <- getBeta(GRSets_norm_filter)
par(mfrow=c(1,2))
densityPlot(mVals, sampGroups=results$sample_type, main="M values", xlab="M values", legend=F)
legend("topleft", legend = levels(factor(results$sample_type)),
       text.col=RColorBrewer::brewer.pal(8,"Dark2"), cex = 0.6)
densityPlot(bVals, sampGroups=results$sample_type, main="Beta values", xlab="Beta values", legend=F)
legend("top", legend = levels(factor(results$sample_type)),
       text.col=RColorBrewer::brewer.pal(8,"Dark2"), cex = 0.6)

annotations <- getAnnotation(GRSets_norm_filter)
 
```

```{r}
## Assessing Whether Samples are Paired or Not:
multiple_inputs <- results %>%
  dplyr::group_by(sample.submitter_id) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::filter(count > 1)
which(results$sample.submitter_id==multiple_inputs$sample.submitter_id)
bVals[,27] <- (bVals[,27]+bVals[,167])/2
bVals[,92] <- (bVals[,92]+bVals[,296])/2
bVals <- bVals[,-c(167,296)]
results <- results[-c(167,296),]
colnames(bVals) <- results$sample.submitter_id

paired_samples <- results %>%
  dplyr::group_by(cases.submitter_id) %>%
  dplyr::mutate(count = n_distinct(sample_type)) %>%
  dplyr::filter(count > 1) %>% dplyr::ungroup() %>%
  dplyr::arrange(cases.submitter_id)

unpaired_controls <- results %>%
  dplyr::filter(sample_type == "Solid Tissue Normal",
         !(cases.submitter_id %in% paired_samples$cases.submitter_id)) %>%
  pull(sample.submitter_id)

unpaired_cases <- results %>%
 dplyr::filter(sample_type == "Primary Tumor",
         !(cases.submitter_id %in% paired_samples$cases.submitter_id)) %>%
  pull(sample.submitter_id)

```

## Paired and Independent T-Tests:

```{r}
# Paired t-tests
paired_cases <- paired_samples %>%
  filter(sample_type == "Primary Tumor") %>%
  pull(sample.submitter_id)

paired_controls <- paired_samples %>%
  filter(sample_type == "Solid Tissue Normal") %>%
  pull(sample.submitter_id)

CpG_Site = rownames(bVals)

paired_t_test_res <- c() # change to a vector, so it will be faster (I guess)
for (j in 1:nrow(bVals)){
    paired_t_test_res_temp = t.test(bVals[j,paired_cases],bVals[j,paired_controls],alternative="two.sided",paired=T)$p.value
    paired_t_test_res <- c(paired_t_test_res, paired_t_test_res_temp)
}
paired_t_test_res = data.frame(cbind(p.value=paired_t_test_res, CpG_Site))

paired_t_test_res_FDR = paired_t_test_res %>%
  mutate(FDR_P_Value = p.adjust(paired_t_test_res$p.value,method="fdr"))

# Two-sample t-tests
ind_cases <- unpaired_cases
ind_controls <- c(unpaired_controls, paired_controls)

t_test_res <- c()
for (j in 1:nrow(bVals)){
    t_test_res_temp = t.test(bVals[j,ind_cases],bVals[j,ind_controls],alternative="two.sided")$p.value
    t_test_res <- c(t_test_res, t_test_res_temp)
}

t_test_res = data.frame(cbind(p.value=t_test_res, CpG_Site))
t_test_res_FDR = t_test_res %>%
  mutate(FDR_P_Value = p.adjust(t_test_res$p.value,method="fdr"))
## End of tests here, can u check if the previous part is correct?
```

## Results:

### Manhattan Plot of FDR Corrected P-Values from Paired T Test:

```{r}
chrom_info = data.frame(rowRanges(GRSets_norm_filter))%>%
  dplyr::select("seqnames","start")

chrom_info = cbind(chrom_info,GRSets_norm_filter@rowRanges@ranges@NAMES)

colnames(chrom_info)[3] = "probeID"

t_test_FDR_w_chrom = merge(paired_t_test_res_FDR,chrom_info,by.x="CpG_Site",by.y="probeID")

colnames(t_test_FDR_w_chrom) = c("CpG_Site","p.value","FDR_P_Value",
                                 "chrom","pos")

t_test_FDR_w_chrom$FDR_P_Value = -log10(t_test_FDR_w_chrom$FDR_P_Value)

t_test_FDR_w_chrom$chrom = gsub("chr","",t_test_FDR_w_chrom$chrom)

t_test_FDR_w_chrom$chrom = recode(t_test_FDR_w_chrom$chrom,"X" = "23","Y"="24","NA"="25")

t_test_FDR_w_chrom$chrom = as.numeric(t_test_FDR_w_chrom$chrom)

plot <- t_test_FDR_w_chrom %>% 
  dplyr::group_by(chrom) %>% 
  dplyr::summarise(chr_len=max(pos)) %>% 
  dplyr::mutate(tot=cumsum(as.numeric(chr_len))-chr_len) %>%
  dplyr::select(-chr_len) %>%
  dplyr::left_join(t_test_FDR_w_chrom, ., by=c("chrom"="chrom")) %>%
  dplyr::arrange(chrom,pos) %>%
  dplyr::mutate(pos_cum=pos+tot)

axisdf = plot %>% dplyr::group_by(chrom) %>% dplyr::summarize(center=( max(pos_cum) + min(pos_cum) ) / 2 )


ggplot(plot, aes(x=pos_cum, y=FDR_P_Value)) +
    geom_point( aes(color=as.factor(chrom)), alpha=0.8, size=1.3) +
    scale_color_manual(values = rep(c("grey", "skyblue"),22))+
    scale_x_continuous( label = axisdf$chrom[seq(1, length(axisdf$chrom), by = 2)], breaks= axisdf$center[seq(1, length(axisdf$center), by = 2)] ) +
    scale_y_continuous(expand = c(0, 0) ) +
    theme_bw() +
    theme( 
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    )+
  xlab("Chromosome")+
  ylab("-log10(FDR Corrected P-value)")
```

From this manhattan plot of the paired t-test results we can see that 

### Manhattan Plot of FDR Corrected P-Values from Independent T Test:

```{r}
chrom_info = data.frame(rowRanges(GRSets_norm_filter))%>%
  dplyr::select("seqnames","start")

chrom_info = cbind(chrom_info,GRSets_norm_filter@rowRanges@ranges@NAMES)

colnames(chrom_info)[3] = "probeID"

t_test_FDR_w_chrom = merge(t_test_res_FDR,chrom_info,by.x="CpG_Site",by.y="probeID")

colnames(t_test_FDR_w_chrom) = c("CpG_Site","p.value","FDR_P_Value",
                                 "chrom","pos")

t_test_FDR_w_chrom$FDR_P_Value = -log10(t_test_FDR_w_chrom$FDR_P_Value)

t_test_FDR_w_chrom$chrom = gsub("chr","",t_test_FDR_w_chrom$chrom)

t_test_FDR_w_chrom$chrom = recode(t_test_FDR_w_chrom$chrom,"X" = "23","Y"="24","NA"="25")

t_test_FDR_w_chrom$chrom = as.numeric(t_test_FDR_w_chrom$chrom)

plot <- t_test_FDR_w_chrom %>% 
  group_by(chrom) %>% 
  summarise(chr_len=max(pos)) %>% 
  mutate(tot=cumsum(as.numeric(chr_len))-chr_len) %>%
  dplyr::select(-chr_len) %>%
  left_join(t_test_FDR_w_chrom, ., by=c("chrom"="chrom")) %>%
  arrange(chrom,pos) %>%
  mutate(pos_cum=pos+tot)

axisdf = plot %>% group_by(chrom) %>% summarize(center=( max(pos_cum) + min(pos_cum) ) / 2 )

axisdf$chrom = c(seq(1,22))

ggplot(plot, aes(x=pos_cum, y=FDR_P_Value)) +
    geom_point( aes(color=as.factor(chrom)), alpha=0.8, size=1.3) +
    scale_color_manual(values = rep(c("grey", "skyblue"),22))+
    scale_x_continuous( label = axisdf$chrom[seq(1, length(axisdf$chrom), by = 2)], breaks= axisdf$center[seq(1, length(axisdf$center), by = 2)] ) +
    scale_y_continuous(expand = c(0, 0) ) +
    theme_bw() +
    theme( 
      legend.position="none",
      panel.border = element_blank(),
      panel.grid.major.x = element_blank(),
      panel.grid.minor.x = element_blank()
    )+
  xlab("Chromosome")+
  ylab("-log10(FDR Corrected P-value)")
```

### Volcano Plot to pinpoint DMR's (Paired Samples):

```{r}
bVals_t = data.frame(t(bVals))

diff_mean_paired_Case = map_dfr(bVals_t[paired_cases,],mean,na.rm=T)
diff_mean_paired_Control = map_dfr(bVals_t[paired_controls,],mean,na.rm=T)
diff_mean_paired = diff_mean_paired_Case-diff_mean_paired_Control
diff_mean_paired = pivot_longer(diff_mean_paired,1:ncol(diff_mean_paired),values_to = "diff_mean", names_to = "CpG_Site")
diff_mean_paired = merge(diff_mean_paired,paired_t_test_res_FDR,by="CpG_Site")
diff_mean_paired$diffexpressed <- "Not Significant"
diff_mean_paired$diffexpressed[diff_mean_paired$diff_mean > 0.25 & diff_mean_paired$FDR_P_Value < 10^-5] <- "Hypermethylated"
diff_mean_paired$diffexpressed[diff_mean_paired$diff_mean < 0.25 & diff_mean_paired$FDR_P_Value < 10^-5] <- "Hypomethylated"
diff_mean_paired$diffexpressed[diff_mean_paired$diff_mean <= 0.25 & diff_mean_paired$diff_mean >= -0.25&diff_mean_paired$FDR_P_Value < 10^-5] <- "No Difference In Methylation"

ggplot(diff_mean_paired, aes(x=diff_mean, y=-log10(FDR_P_Value),col=diffexpressed)) +
    geom_point(alpha=0.8, size=1.3)+
    theme_bw() +
  scale_color_manual(values=c("black", "blue", "pink","red"))+
    theme(legend.position="top")+
  guides(fill=guide_legend(title="Methylation (Case vs Controls)"))+
  geom_hline(yintercept = 5, color = "darkblue")+
  geom_vline(xintercept = 0.25,color="darkred")+
  geom_vline(xintercept = -0.25,color="darkred")+
  xlab("Difference in Mean Methylation between Case and Controls")+
  ylab("-log10(FDR Corrected P-value)")+
  ggtitle("Volcano Plot showing DMR's for Paired Case/Control Samples")
```

### Volcano Plot to pinpoint DMR's (Independent Samples):

```{r}
diff_mean_Idep_Case = map_dfr(bVals_t[ind_cases,],mean,na.rm=T)
diff_mean_Idep_Control = map_dfr(bVals_t[ind_controls,],mean,na.rm=T)
diff_mean_Idep = diff_mean_Idep_Case-diff_mean_Idep_Control
diff_mean_Idep = pivot_longer(diff_mean_Idep,1:ncol(diff_mean_Idep),values_to = "diff_mean", names_to = "CpG_Site")
diff_mean_Idep = merge(diff_mean_Idep,t_test_res_FDR,by="CpG_Site")
diff_mean_Idep$diffexpressed <- "Not Significant"
diff_mean_Idep$diffexpressed[diff_mean_Idep$diff_mean > 0.25 & diff_mean_Idep$FDR_P_Value < 10^-5] <- "Hypermethylated"
diff_mean_Idep$diffexpressed[diff_mean_Idep$diff_mean < 0.25 & diff_mean_Idep$FDR_P_Value < 10^-5] <- "Hypomethylated"
diff_mean_Idep$diffexpressed[diff_mean_Idep$diff_mean <= 0.25 & diff_mean_Idep$diff_mean >= -0.25&diff_mean_Idep$FDR_P_Value < 10^-5] <- "No Difference In Methylation"

ggplot(diff_mean_Idep, aes(x=diff_mean, y=-log10(FDR_P_Value),col=diffexpressed)) +
    geom_point(alpha=0.8, size=1.3)+
    theme_bw() +
  scale_color_manual(values=c("black", "blue", "pink","red"))+
    theme(legend.position="top")+
  guides(fill=guide_legend(title="Methylation (Case vs Controls)"))+
  geom_hline(yintercept = 5, color = "darkblue")+
  geom_vline(xintercept = 0.25,color="darkred")+
  geom_vline(xintercept = -0.25,color="darkred")+
  xlab("Difference in Mean Methylation between Case and Controls")+
  ylab("-log10(FDR Corrected P-value)")+
  ggtitle("Volcano Plot showing DMR's for Independent Case/Control Samples")
```

### Overlap between Paired and Independent Tests:


## Discussion and Conclusions:
